[Frontend] Tool calling parser for Granite 3.0 models

maxdebayser opened the pull request:
This PR adds a tool calling parser for `ibm-granite/granite-3.0-8b-instruct`. The smaller models in 
the [Granite 3.0 Language Models](https://huggingface.co/collections/ibm-granite/granite-30-models-66fdb59bbb54785c3512114f) models collection pass some of the tests..

cc: @njhill

njhill commented on docs/source/serving/openai_compatible_server.md:
```diff
@@ -157,7 +157,7 @@ vLLM will use guided decoding to ensure the response matches the tool parameter
 * `--enable-auto-tool-choice` -- **mandatory** Auto tool choice. tells vLLM that you want to enable the model to generate its own tool calls when it 
 deems appropriate.
-* `--tool-call-parser` -- select the tool parser to use - currently either `hermes` or `mistral` or `llama3_json` or `internlm`. Additional tool parsers 
+* `--tool-call-parser` -- select the tool parser to use - currently either `hermes`, `mistral`, `llama3_json`, `internlm` or `granite-20b-fc`. Additional tool parsers 
```
@maxdebayser should `granite` also be added here?

njhill commented on docs/source/serving/openai_compatible_server.md:
```diff
@@ -219,6 +222,20 @@ it works better with vLLM.
+Flags: `--tool-call-parser granite-20b-fc`
+`examples/tool_chat_template_granite_20b_fc.jinja`: this is a modified chat template from the original on Huggingface, which is not vLLM compatible. It blends function description elements from the Hermes template and follows the same system prompt as "Response Generation" mode from [the paper](https://arxiv.org/abs/2407.00121). Parallel function calls are supported.
+
+* `ibm-granite/granite-8b-instruct`
```
```suggestion
* `ibm-granite/granite-3.0-8b-instruct`
```

njhill commented on docs/source/serving/openai_compatible_server.md:
```diff
@@ -219,6 +222,20 @@ it works better with vLLM.
+* `ibm-granite/granite-20b-functioncalling`
+
+Flags: `--tool-call-parser granite-20b-fc`
+`examples/tool_chat_template_granite_20b_fc.jinja`: this is a modified chat template from the original on Huggingface, which is not vLLM compatible. It blends function description elements from the Hermes template and follows the same system prompt as "Response Generation" mode from [the paper](https://arxiv.org/abs/2407.00121). Parallel function calls are supported.
```
Is `--chat-template` missing here? (see llama above for example)

njhill commented on docs/source/serving/openai_compatible_server.md:
```diff
@@ -219,6 +222,20 @@ it works better with vLLM.
+* `ibm-granite/granite-8b-instruct`
+
+Flags: `--tool-call-parser granite`
+`examples/tool_chat_template_granite.jinja`: this is a modified chat template from the original on Huggingface. Parallel function calls are supported.
```
Is `--chat-template` missing here?

njhill commented on docs/source/serving/openai_compatible_server.md:
```diff
@@ -230,6 +247,15 @@ Known issues:
 Recommended flags: `--tool-call-parser internlm --chat-template examples/tool_chat_template_internlm2_tool.jinja`
 
 
+#### IBM Granite
+
+Supported models:
+* `ibm-granite/granite-20b-functioncalling`
+
+Flags: `--tool-call-parser granite-20b-fc`
+`examples/tool_chat_template_granite_20b_fc.jinja`: this is a modified chat template from the original on Huggingface, which is not vLLM compatible. It blends function description elements from the Hermes template and follows the same system prompt as "Response Generation" mode from [the paper](https://arxiv.org/abs/2407.00121). Parallel function calls are supported.
+
+
```
Duplicate, should be removed?

njhill commented on docs/source/serving/openai_compatible_server.md:
```diff
@@ -219,6 +222,20 @@ it works better with vLLM.
+#### IBM Granite
+
+Supported models:
+* `ibm-granite/granite-20b-functioncalling`
```
Probably better to swap the order of these and move the granite 3.0 one to the top

njhill reviewed (commented):
Thanks @maxdebayser!

maxdebayser committed:
Add tool calling support for granite-3.0-8b-instruct

This model supports all the cases in our unit tests

I had to rebase this due to DCO problems in several commits
that have now been merged in main.

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
lower max model len for tool calling tests

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
revert forcing fast check

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser commented:
This is ready for review.

cc: @njhill @wseaton

njhill commented on vllm/entrypoints/openai/tool_parsers/granite_tool_parser.py:
```diff
@@ -0,0 +1,227 @@
+                    elif prev_arguments:
+                        prev_args_json = json.dumps(prev_arguments)
+                        if cur_args_json != prev_args_json:
+
```
nit: remove this empty line?

njhill reviewed (commented):
Thanks @maxdebayser! I did a first pass and left some comments

njhill commented:
cc @K-Mistele in case you'd like to take a look too

maxdebayser committed:
add review suggestions

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
add review suggestions

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
add review suggestions

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
fix linting

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

K-Mistele commented:
will take another look!

njhill reviewed (approved):
Thanks again @maxdebayser. I'll hold off for any comments from @K-Mistele before merging.

K-Mistele commented:
Giving it a final once-over :) 

Out of curiosity it looks like the model's default context is 4096; is there any way to scale this if I want to test longer-context tool calls? no worries if there's not a good way to do with with vLLM

K-Mistele commented:
Looks great to me! Seems pretty robust, and passing tests on my machine :)

K-Mistele commented:
Looks like the only failing tests were AMD-related which was expected last I checked. cc @mgoin @DarkLight1337

DarkLight1337 commented:
> Looks like the only failing tests were AMD-related which was expected last I checked. cc @mgoin @DarkLight1337

The test is not failing on main branch, so I think it's introduced by this PR. Do tell me if I'm mistaken though.

K-Mistele commented:
> > Looks like the only failing tests were AMD-related which was expected last I checked. cc @mgoin @DarkLight1337
> 
> The test is not failing on main branch, so I think it's introduced by this PR. Do tell me if I'm mistaken though.

Oops, missed that. I just remembered the last time I did one there were some expected to fail, but I'm sure you're right :) 

Seems related to the FP8 granite 20B quantization (`mbayser/granite-20b-functioncalling-FP8-KV`) that's being used:

https://buildkite.com/vllm/ci-aws/builds/10840#01930511-9576-45cb-aebb-e237d5f07c9b/974-5501

AMD supports FP8 in ROCM > 6.2. Could also be an OOM issue since this is a 20B that we had issues fitting into other CI GPUs; or could be related to CPU offloading. 

Whatever the case, seems more like a model+configuration+hardware compatibility issue rather than a code issue. Is there any way to disable this particular model for AMD tests, and would that be an acceptable solution?

DarkLight1337 commented:
I think this is alright. @njhill any thoughts?

maxdebayser committed:
Merge branch 'upstream_main' into granite-8b-instruct-2

maxdebayser committed:
skip quantized granite-20b test on ROCm

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser commented:
I've pushed a change to disable skip the `granite-20b-functioncalling` test on AMD. Let's see if it passes now :crossed_fingers:

maxdebayser commented:
> Out of curiosity it looks like the model's default context is 4096; is there any way to scale this if I want to test longer-context tool calls? no worries if there's not a good way to do with with vLLM

I'm not sure, but as we add more models and tests, we might have to get the max sequence length for each model and skip the tests that require longer ones.

maxdebayser committed:
fix copy&paste error :facepalm

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

njhill commented:
Thanks again @maxdebayser @K-Mistele @DarkLight1337!
